plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,250 @@ extern "C"
 
-        timeout = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+        timeout_dpms = wlr_idle_timeout_create(wf::get_core().protocols.idle,
             wf::get_core().get_current_seat(), 1000 * timeout_sec);
 
-        on_idle.set_callback([&] (void*) {
+        on_idle_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_SELF, wf::OUTPUT_IMAGE_SOURCE_DPMS);
         });
-        on_idle.connect(&timeout->events.idle);
+        on_idle_dpms.connect(&timeout_dpms->events.idle);
 
-        on_resume.set_callback([&] (void*) {
+        on_resume_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_DPMS, wf::OUTPUT_IMAGE_SOURCE_SELF);
         });
-        on_resume.connect(&timeout->events.resume);
+        on_resume_dpms.connect(&timeout_dpms->events.resume);
+    }
+
+    void create_screensaver_timeout(int timeout_sec)
+    {
+        destroy_screensaver_timeout();
+        if (timeout_sec <= 0)
+            return;
+
+        timeout_screensaver = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+            wf::get_core().get_current_seat(), 1000 * timeout_sec);
+
+        on_idle_screensaver.set_callback([&] (void*) {
+            start_screensaver();
+        });
+        on_idle_screensaver.connect(&timeout_screensaver->events.idle);
+
+        on_resume_screensaver.set_callback([&] (void*) {
+            stop_screensaver();
+        });
+        on_resume_screensaver.connect(&timeout_screensaver->events.resume);
+    }
+
+    void screensaver_frame()
+    {
+        cube_control_signal data;
+        bool carried_out = true;
+
+        if (screensaver_stopping)
+            rotation = duration.progress(rot_transition);
+        else
+            rotation += cube_rotate_speed->as_double();
+        if (rotation >= M_PI * 2)
+            rotation = 0.0;
+
+        data.angle = rotation;
+        data.zoom = duration.progress(zoom_transition);
+        data.last_frame = false;
+
+        if (screensaver_stopping && !duration.running())
+        {
+            cube_control_signal data;
+            data.angle = 0.0;
+            data.zoom = ZOOM_BASE;
+            data.last_frame = true;
+            for (auto& output : wf::get_core().output_layout->get_outputs()) {
+                output->emit_signal("cube-control", &data);
+                if (data.carried_out) {
+                    output->render->rem_effect(&render_hook);
+                }
+                if (outputs_inhibited) {
+                    output->render->add_inhibit(false);
+                    output->render->damage_whole();
+                }
+            }
+            outputs_inhibited = false;
+            screensaver_stopping = false;
+            return;
+        }
+
+        for (auto& output : wf::get_core().output_layout->get_outputs()) {
+            output->emit_signal("cube-control", &data);
+            if (!data.carried_out) {
+                if (!outputs_inhibited) {
```

Comment:
Okay so we have a problem here. Cube may fail on some outputs, but not on the others. Which means some output will get inhibited, others not.

Then upon resume, you will try to add_inhibit(false) to all outputs (because outputs_inhibitied is `true`) but it is not true for all outputs.

What I'd suggest: first iterate over all outputs, and try to activate cube. If you can't activate it for some output, just iterate over all outputs again, disable cube on outputs where it was activated, then just turn off all outputs.